components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/CacheBackedApplicationDAO.java

```diff
@@ -0,0 +1,315 @@
+ *  under the License.
+ */
+
+package org.wso2.carbon.identity.application.mgt.dao.impl;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.context.CarbonContext;
+import org.wso2.carbon.identity.application.common.IdentityApplicationManagementException;
+import org.wso2.carbon.identity.application.common.model.ApplicationBasicInfo;
+import org.wso2.carbon.identity.application.common.model.ClaimConfig;
+import org.wso2.carbon.identity.application.common.model.ClaimMapping;
+import org.wso2.carbon.identity.application.common.model.InboundAuthenticationRequestConfig;
+import org.wso2.carbon.identity.application.common.model.ServiceProvider;
+import org.wso2.carbon.identity.application.mgt.ApplicationMgtUtil;
+import org.wso2.carbon.identity.application.mgt.cache.IdentityServiceProviderCache;
+import org.wso2.carbon.identity.application.mgt.cache.IdentityServiceProviderCacheEntry;
+import org.wso2.carbon.identity.application.mgt.cache.IdentityServiceProviderCacheKey;
+import org.wso2.carbon.identity.application.mgt.dao.ApplicationDAO;
+import org.wso2.carbon.identity.application.mgt.internal.cache.ServiceProvideCacheInboundAuth;
+import org.wso2.carbon.identity.application.mgt.internal.cache.ServiceProvideCacheInboundAuthEntry;
+import org.wso2.carbon.identity.application.mgt.internal.cache.ServiceProvideCacheInboundAuthKey;
+import org.wso2.carbon.identity.application.mgt.internal.cache.ServiceProviderCacheID;
+import org.wso2.carbon.identity.application.mgt.internal.cache.ServiceProviderIDCacheEntry;
+import org.wso2.carbon.identity.application.mgt.internal.cache.ServiceProviderIDCacheKey;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Cached DAO layer for the application management. All the DAO access has to be happen through this layer to ensure
+ * single point of caching.
+ */
+public class CacheBackedApplicationDAO extends AbstractApplicationDAOImpl {
+
+    private static final Log log = LogFactory.getLog(CacheBackedApplicationDAO.class);
+
+    private ApplicationDAO appDAO;
+
+    private static IdentityServiceProviderCache appCacheByName = null;
+    private static ServiceProvideCacheInboundAuth appCacheByInboundAuth = null;
+    private static ServiceProviderCacheID appCacheByID = null;
+
+    public CacheBackedApplicationDAO(ApplicationDAO appDAO) {
+
+        this.appDAO = appDAO;
+        appCacheByName = IdentityServiceProviderCache.getInstance();
+        appCacheByInboundAuth = ServiceProvideCacheInboundAuth.getInstance();
+        appCacheByID = ServiceProviderCacheID.getInstance();
+    }
+
+    public ServiceProvider getApplication(String applicationName, String tenantDomain) throws
+            IdentityApplicationManagementException {
+
+        ServiceProvider serviceProvider = getApplicationFromCache(applicationName, tenantDomain);
```

Comment:
add debug log to indicate a cache hit or a miss

Reply:
added with https://github.com/wso2/carbon-identity-framework/pull/2112/commits/028b538c9deb0987b027a77d46f237652f4eeabc